test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Tests.ps1

```diff
@@ -0,0 +1,247 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+$script:oldModulePath = $env:PSModulePath
+
+function Add-ModulePath
+{
+    param([string]$Path)
+
+    $script:oldModulePath = $env:PSModulePath
+
+    $env:PSModulePath = $env:PSModulePath + [System.IO.Path]::PathSeparator + $Path
+}
+
+function Restore-ModulePath
+{
+    $env:PSModulePath = $script:oldModulePath
+}
+
+# Creates a new dummy module compatible with the given PSEditions
+function New-EditionCompatibleModule
+{
+    param(
+        [Parameter(Mandatory = $true)][string]$ModuleName,
+        [string]$DirPath,
+        [string[]]$CompatiblePSEditions)
+
+    $modulePath = Join-Path $DirPath $ModuleName
+
+    $manifestPath = Join-Path $modulePath "$ModuleName.psd1"
+
+    $psm1Name = "$ModuleName.psm1"
+    $psm1Path = Join-Path $modulePath $psm1Name
+
+    New-Item -Path $modulePath -ItemType Directory
+
+    New-Item -Path $psm1Path -Value "function Test-$ModuleName { `$true }"
+
+    if ($CompatiblePSEditions)
+    {
+        New-ModuleManifest -Path $manifestPath -CompatiblePSEditions $CompatiblePSEditions -RootModule $psm1Name
+    }
+    else
+    {
+        New-ModuleManifest -Path $manifestPath -RootModule $psm1Name
+    }
+
+    return $modulePath
+}
+
+function New-TestModules
+{
+    param([hashtable[]]$TestCases, [string]$BaseDir)
+
+    for ($i = 0; $i -lt $TestCases.Count; $i++)
+    {
+        $path = New-EditionCompatibleModule -ModuleName $TestCases[$i].ModuleName -CompatiblePSEditions $TestCases[$i].Editions -Dir $BaseDir
+
+        $TestCases[$i].Path = $path
+        $TestCases[$i].Name = $TestCases[$i].Editions -join ","
+    }
+}
+
+Describe "Get-Module with CompatiblePSEditions-checked paths" -Tag "CI" {
+
+    BeforeAll {
+        $successCases = @(
+            @{ Editions = "Core","Desktop"; ModuleName = "BothModule" },
+            @{ Editions = "Core"; ModuleName = "CoreModule" }
+        )
+
+        $failCases = @(
+            @{ Editions = "Desktop"; ModuleName = "DesktopModule" },
+            @{ Editions = $null; ModuleName = "NeitherModule" }
+        )
+
+        $basePath = Join-Path $TestDrive "EditionCompatibleModules"
+        New-TestModules -TestCases $successCases -BaseDir $basePath
+        New-TestModules -TestCases $failCases -BaseDir $basePath
+
+        # Emulate the System32 module path for tests
+        [System.Management.Automation.Internal.InternalTestHooks]::SetTestHook("TestWindowsPowerShellPSHomeLocation", $basePath)
+    }
+
+    AfterAll {
+        Remove-Item -Force -Recurse $basePath
```

Comment:
You don't have to remove the modules as Pester will automatically clean out the `$TestDrive`.